### DISCUSSION TITLE
How to get the type of a variable?

### DISCUSSION THREAD

**Discussion by rempas** [MrSmith33/vox#25] [Q&A] [answered]
How can I use a compile time if (#if) to mach the type of a variable? In D, I would had the following code:

```
void print_type(T)(T arg) {
  static if (is(typeof(arg) == string))
    write("It is a string");
  else static if (is(typeof(arg) == int))
    write("Is is an int");
}
```

Of course I know that other stuff are different in Vox but what I'm interested is the conditional code for the `if` and `else if` statement

**Comment by MrSmith33** [comment, marked as answer]
I didn't implement `typeof` yet, but in that concrete example you could use `T`

**Reply by rempas** [reply]
Alright! It works great! If that work, then is `typeof` needed anyway? Also what's the type of a string literal that is passed to this function? The following code doesn't match anything:

```
enum u32 stdin  = 0;
enum u32 stdout = 1;
enum u32 stderr = 2;

@extern(syscall, 60)
void exit(i32 error_code);

@extern(syscall, 1)
void sys_write(u32 fd, u8* buf, u64 count);

void write(u32 fd, u8[] data) {
  sys_write(fd, data.ptr, data.length);
}

void print_type[T](T arg) {
  #if (T == i8*)
    write(1, "It's i8*\n");
  else #if (T == u8*)
    write(1, "It's u8*\n");
  else #if (T == i16*)
    write(1, "It's i16*\n");
  else #if (T == u16*)
    write(1, "It's u16*\n");
  else #if (T == i32*)
    write(1, "It's i32*\n");
  else #if (T == u32*)
    write(1, "It's u32*\n");
  else #if (T == i64*)
    write(1, "It's i64*\n");
  else #if (T == u64*)
    write(1, "It's u64*\n");
  else #if (T == i8[])
    write(1, "It's i8[]\n");
  else #if (T == u8[])
    write(1, "It's u8[]\n");
  else #if (T == i16[])
    write(1, "It's i16[]\n");
  else #if (T == u16[])
    write(1, "It's u16[]\n");
  else #if (T == i32[])
    write(1, "It's i32[]\n");
  else #if (T == u32[])
    write(1, "It's u32[]\n");
  else #if (T == i64[])
    write(1, "It's i64[]\n");
  else #if (T == u64[])
    write(1, "It's u64[]\n");
  else
    write(1, "Another type?\n");
}

void main() {
  print_type(21);
  print_type("Hello!!!");
  exit(0);
}
```

**Reply by MrSmith33** [reply]
ah yes, there is a bit of a problem with generic types in Vox (things like `T*, T[], T[n]`). They are not yet deduplicated, so comparing them with `==` doesn't yet work, because lhs and rhs point to different instances of those types. It works however for basic types and user-defined types. For those you can use `$isSlice, $isInteger, $isPointer, $baseOf` built-in functions. <https://github.com/MrSmith33/voxelman2/blob/master/plugins/core/utils.vx#L203>

**Reply by rempas** [reply]
Thanks! Do you plan to change this behavior some time in the future so we can compare them? Also why have you choose to make string literals get passed as slices rather than simple pointers who are faster? Is there a particular advantage?

**Reply by MrSmith33** [reply]
Yea, I'm thinking about type deduplication, but it isn't an easy thing to implement. So may take a while.

String literals are zero terminated, and they implicitly cast to `u8*` (useful for C interop). Also, you can get the raw pointer with `.ptr` property if you really want that. Otherwise strings are just regular slices and knowing the length is always a good idea.

**Reply by rempas** [reply]
> Yea, I'm thinking about type deduplication, but it isn't an easy thing to implement. So may take a while

That sounds awesome! Of course it will take a lot of time to implement that but who cares really? Good things take time so it makes sense!

Now about strings. Are slices Objects or special types that are handled by the compiler. My point being, when the implicit conversion happens, will this be just as taking the `.ptr` part of the slice or will we need to have instructions to make a conversion at compile time? Or is it a compile time things that you only play with the memory that is used? I'm so confused...

**Reply by MrSmith33** [reply]
Slices are special types, but they are effectively a struct with 2 members `ulong length; T* ptr`. So string literal is a struct literal with length and pointer to static data. The implicit conversion of string to `char*` is the same as using `.ptr` on the literal. See [spec/index.md#slices](https://github.com/MrSmith33/vox/blob/master/spec/index.md#slices)

**Reply by rempas** [reply]
Cool! That's nice! Thanks a lot for you time man!
